peps/pep-0733.rst

```diff
@@ -0,0 +1,683 @@
+
+C API problems
+==============
+
+The remainder of this document summarizes and categorizes the problems that were reported on
+the `capi-workgroup <https://github.com/capi-workgroup/problems/issues/>`__ repository.
+The issues are grouped into several categories.
+
+
+API Evolution and Maintenance
+-----------------------------
+
+The difficulty of making changes in the C API is central to this report. It is
+implicit in many of the issues we discuss here, particularly when we need to
+decide whether an incremental bugfix can resolve the issue, or whether it can
+only be addressed as part of an API redesign
+[`Issue 44 <https://github.com/capi-workgroup/problems/issues/44>`__]. The
+benefit of each incremental change is often viewed as too small to justify the
+disruption. Over time, this implies that every mistake we make in an API's
+design or implementation remains with us indefinitely.
+
+We can take two views on this issue. One is that this is a problem and the
+solution needs to be baked into any new C API we design, in the form of a
+process for incremental API evolution, which includes deprecation and
+removal of API elements. The other possible approach is that this is not
+a problem to be solved, but rather a feature of any API. In this
+view, API evolution should not be incremental, but rather through large
+redesigns, each of which learns from the mistakes of the past and is not
+shackled by backwards compatibility requirements (in the meantime, new
+API elements may be added, but nothing can ever be removed). A compromise
+approach is somewhere between these two extremes, fixing issues which are
+easy or important enough to tackle incrementally, and leaving others alone.
+
+The problem we have in CPython is that we don't have an agreed, official
+approach to API evolution. Different members of the core team are pulling in
+different directions and this is an ongoing source of disagreements.
+Any new C API needs to come with a clear decision about the model
+that its maintenance will follow, as well as the technical and
+organizational processes by which this will work.
+
+If the model does include provisions for incremental evolution of the API,
+it will include processes for managing the impact of the change on users
+[`Issue 60 <https://github.com/capi-workgroup/problems/issues/60>`__],
+perhaps through introducing an external backwards compatibility module
+[`Issue 62 <https://github.com/capi-workgroup/problems/issues/62>`__],
+or a new API tier of "blessed" functions
+[`Issue 55 <https://github.com/capi-workgroup/problems/issues/55>`__].
+
+
+API Specification and Abstraction
+---------------------------------
+
+The C API does not have a formal specification, it is described
+semi-formally in the documentation and exposed through C header
+files. This is also true for the limited API and the stable ABI.
+This creates a number of problems.
```

Comment:
This is a nitpick, but the limited API and stable ABI *do* have a machine-readable representation for external tools to consume: <https://github.com/python/cpython/blob/main/Misc/stable_abi.toml>. It might be worth pointing that out, as a starting point or source of inspiration for improving the overall specify-ability of the C API.

(I use that representation to generate `abi3info`, which in turn powers tools like `abi3audit`.)

Reply:
Since we're focused here on problems rather than solutions, I think this could be worth mentioning on the relevant issue, but it wouldn't add clarity to this section (which currently doesn't mention the stable ABI or limited API at all).

Reply:
The point about solutions is a fair one 🙂 

>  but it wouldn't add clarity to this section (which currently doesn't mention the stable ABI or limited API at all).

Forgive me if I'm misunderstanding, but is this true? Here's the part of the paragram that I honed in on:

> This is also true for the limited API and the stable ABI.

My interpretation of that sentence is that it extends the claim about the lack of a formal specification to the limited API and stable ABI. But perhaps that isn't what was intended?

Reply:
Ah right, sorry. Let me think about how this could be corrected.

Reply:
"Formal specification" isn't the same as "machine readable representation". The header files are technically machine readable representations, but that doesn't mean the whole interpreter is formally specified 😉 

The specification is currently the reference interpreter, when it should be a written document full of "SHOULDs" and "SHALLs" and stuff like that. How about something like:

"The C API does not have a formal specification, but is currently defined as whatever the reference implementation (CPython) defines in a particular version. The documentation acts as an incomplete description, but is not sufficient for verifying the correctness of either the full API, the limited API, or the stable ABI. As a result, the C API may change significantly between releases without needing a more visible specification update, which leads to a number of problems." (I assume you go on to describe these problems, otherwise the last bit is a bit strange, but I don't want to do the seven clicks to get GitHub to expand it without losing my comment right now.)

Reply:
> "Formal specification" isn't the same as "machine readable representation". The header files are technically machine readable representations, but that doesn't mean the whole interpreter is formally specified 😉

Yep, this is why I called it a nitpick and said "machine representation." I thought it would possibly be worth noting that there *are* other (very limited!) descriptions of the C API besides the source code itself.

But anybody who is so motivated will be able to discover those anyways, so I don't feel strongly about this 🙂